.github/workflows/snapshot-build.yml

```diff
@@ -123,16 +123,40 @@ jobs:
         uses: actions-rs/cargo@v1
         with:
           command: build
-          args: --profile production --bins --features=subspace-node/json-chain-spec,subspace-farmer/cuda
+          args: --profile production --bin subspace-node --bin subspace-farmer --features=subspace-node/json-chain-spec,subspace-farmer/cuda
         if: runner.os == 'Linux' || runner.os == 'Windows'
 
       - name: Build (macOS without CUDA)
         uses: actions-rs/cargo@v1
         with:
           command: build
-          args: --profile production --bins --features=subspace-node/json-chain-spec
+          args: --profile production --bin subspace-node --bin subspace-farmer --features=subspace-node/json-chain-spec
         if: runner.os == 'macOS'
-
+      # macOS Code Signing
+      - name: Sign Application (macOS)
+        run: |
+          echo "${{ secrets.MACOS_CERTIFICATE }}" | base64 --decode > certificate.p12
+          security create-keychain -p "${{ secrets.MACOS_CERTIFICATE_PW }}" defaultKeychain build.keychain
+          security default-keychain -s build.keychain
+          security unlock-keychain -p "${{ secrets.MACOS_CERTIFICATE_PW }}" build.keychain
+          security import certificate.p12 -k build.keychain -P "${{ secrets.MACOS_CERTIFICATE_PW }}" -T /usr/bin/codesign
+          security set-key-partition-list -S apple-tool:,apple:,codesign: -s -k "${{ secrets.MACOS_CERTIFICATE_PW }}" build.keychain
+          /usr/bin/codesign --force --options=runtime -s "${{ secrets.MACOS_IDENTITY }}" --timestamp target/production/subspace-node
+          /usr/bin/codesign --force --options=runtime -s "${{ secrets.MACOS_IDENTITY }}" --timestamp target/production/subspace-farmer
+          /usr/bin/xcrun altool --notarize-app --primary-bundle-id node-${{ github.ref_name }} --username "${{ secrets.MACOS_APPLE_ID}}" --password "${{ secrets.MACOS_APP_PW }}" --file target/production/subspace-node
+          /usr/bin/xcrun altool --notarize-app --primary-bundle-id farmer-${{ github.ref_name }} --username "${{ secrets.MACOS_APPLE_ID}}" --password "${{ secrets.MACOS_APP_PW }}" --file target/production/subspace-farmer
+        if: runner.os == 'macOS'
+      # Windows Code Signing
+      - name: Sign Application (Windows)
+        uses: Dana-Prajea/code-sign-action@v4
+        with:
+          certificate: '${{ secrets.WINDOWS_CERTIFICATE }}'
+          password: '${{ secrets.WINDOWS_CERTIFICATE_PW }}'
+          certificatesha1: '00A427587B911908F59B6C42BA2863109C599C1C'
+          folder: 'target/production'
+          recursive: false
+          timestampUrl: http://timestamp.comodoca.com
+        if: runner.os == 'Windows'
       - name: Prepare executables for uploading (Linux or macOS)
         run: |
           mkdir executables
```

Comment:
In the error log, it says macOS executables should have the extension **zip, pkg or dmg** (https://github.com/subspace/subspace/runs/5434823346?check_suite_focus=true)

I guess we should separate linux and mac executable preparation in here, and add the extension of `.dmg` for macOS

Reply:
probably something like:
`mv target/production/subspace-node executables/subspace-node-x86_64-${{ matrix.os }}-${{ github.ref_name }}.dmg` 
for the macOS build

Reply:
You can't just rename random file to dmg and expect it to work.

Here is a DM I sent to Justin yesterday:
> Now we have a different problem with code signing on macOS: https://github.com/subspace/subspace/runs/5434823346?check_suite_focus=true
Looks like you can't send just a binary for notarization, you need to send a bundle or something like that.
Apparently it is possible to just zip signed files and send for notarization, but I'm not sure if we can still distribute just executable.
Can you try to sign one of the binaries from Janyary 5th, send it for notarization, wait for that to finish and see if you can run executable without any warnings and special actions?
See "Code-signing", "Zip it up for notarization" and "Notarization" steps in https://glyphsapp.com/learn/how-to-notarize-your-plug-ins

Reply:
I see, but it's not a random file, it's what rust built for macOS, right? Why can't we just add the `.dmg` extension? It is already in that format, just lacking the explicit extension to my knowledge. I might be wrong though :)

Reply:
Well, why can't you add `.mp3` to it then? The answer is the same :slightly_smiling_face:

Reply:
Because rust built an executable, not a music file? Isn't this file the product of the `build` command?

Reply:
DMG is not an executable either

Reply:
To conclude this, zipping the binary and notarizing works, we will zip each binary and upload them to the release.

Reply:
We will upload original unzipped files to release just like before